src/bpf/bpfd-manager.c

```diff
@@ -0,0 +1,489 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#include <sys/wait.h>
+
+#include "sd-daemon.h"
+#include "sd-messages.h"
+
+#include "bpf-dlopen.h"
+#include "bpfd-conf.h"
+#include "bpfd-manager.h"
+#include "fd-util.h"
+#include "fileio.h"
+#include "fs-util.h"
+#include "hashmap.h"
+#include "missing_syscall.h"
+#include "mkdir.h"
+#include "process-util.h"
+#include "signal-util.h"
+#include "socket-util.h"
+#include "stdio-util.h"
+#include "umask-util.h"
+
+static int get_memlock_by_fd(int fd, char **ret_memlock) {
+        _cleanup_free_ char *fdinfo = NULL;
+        char *memlock, *memlock_line;
+        char path[STRLEN("/proc/self/fdinfo/") + DECIMAL_STR_MAX(int)];
+        int r;
+
+        xsprintf(path, "/proc/self/fdinfo/%i", fd);
+        r = read_full_virtual_file(path, &fdinfo, NULL);
+        if (r < 0)
+                return r;
+
+        memlock_line = find_line_startswith(fdinfo, "memlock:");
+        if (!memlock_line)
+                return -EOPNOTSUPP;
+
+        memlock_line += strspn(memlock_line, WHITESPACE);
+        memlock_line[strcspn(memlock_line, WHITESPACE)] = 0;
+
+        memlock = strdup(memlock_line);
+        if (!memlock)
+                return -ENOMEM;
+
+        *ret_memlock = memlock;
+
+        return 0;
+}
+
+static int fetch_bpf_maps_and_progs(Varlink *link) {
+        int i = 0;
+        uint32_t id = 0;
+        struct bpf_map_info map_info = {};
+        uint32_t map_info_len = sizeof(map_info);
+        struct bpf_prog_info prog_info = {};
+        uint32_t prog_info_len = sizeof(prog_info);
+        uint32_t nr_map_ids = 0;
+        struct bpf_func_info func_info = {};
+        uint32_t func_info_rec_size = 0;
+        const struct btf_type *t;
+        struct btf *btf = NULL;
+        char prog_tag[BPF_TAG_SIZE * 2 + 1];
+        _cleanup_hashmap_free_ Hashmap *map_name_by_id = NULL;
+        int r;
+
+        map_name_by_id = hashmap_new(NULL);
+        if (!map_name_by_id)
+                return -ENOMEM;
+        /* Iterate through all BPF maps to get map names, map types, and memlock */
+        while (sym_bpf_map_get_next_id(id, &id) >= 0) {
+                _cleanup_close_ int fd = -EBADF;
+                _cleanup_free_ char *memlock = NULL;
+                char *map_name = NULL;
+                uint32_t map_id = id;
+                JsonVariant *v;
+
+                zero(map_info);
+                /* Grab the FD using the ID*/
+                fd = sym_bpf_map_get_fd_by_id(id);
+                if (fd < 0) {
+                        log_error("Can't get fd for map id: %u", id);
```

Comment:
Always include the specific error when some operation fails (I'm not 100% sure how libbpf communicates its errors). Also, if you ignore an error, it should be something like `log_error_errno(fd, "Failed to get fd for map id, ignoring: %m");`